addon-test-support/@ember/test-helpers/settled.ts

```diff
@@ -142,12 +143,52 @@ function checkWaiters() {
   return false;
 }
 
+interface IPauseTestPromiseDetails {
+  label: undefined | string;
+  error: Error;
+}
+
+let currentPauseTestPromises = new Map<Promise<unknown>, IPauseTestPromiseDetails>();
+
+/**
+  Ensures that tests wait for the provided promise before being considered "settled".
+
+  This function is expected to be used in application/addon code (aka non-test code).
+  When used during a test run, it ensures that the next `await settled()` check does
+  not result until after the promise has settled, when used outside of a test it is a
+  simple "pass through" function.
+
+  @public
+  @param {Promise<T>} promise the promise to wait for
+  @param {string} [label] an optional label for the promise
+  @returns {Promise<T>} the passed in promise
+*/
+export function pauseTestFor<T>(promise: Promise<T>, label?: string): Promise<T> {
+  let error = new Error();
+
+  currentPauseTestPromises.set(promise, { label, error });
```

Comment:
I guess we have to worry about the case where the same promise is passed in twice, e.g.

```javascript
// my-addon/addon/foo.js
myAddonFunction() {
  let promise = doSomethingAsync();
  pauseTestFor(promise);
  return promise;
}

// my-app/app/bar.js
function doThing() {
  return pauseTestFor(myAddonFunction());
}
```

I know the API is set up to encourage the addon function to `return pauseTestFor(doSomethingAsync())` which would then return a different promise, but it will always be possible to pass the same promise to two different `pauseTestFor()` calls.

I guess doing so wouldn't break the settled behavior -- things would still settle at the same time. The second call would clobber the first call's stack and label, which could potentially be confusing. Alternatively, we could run a `.has()` check and early out and/or issue a warning. It seems like overkill to implement the fully tolerant solution of having the `Set` store an array of `{ label, error }`...thoughts?

Reply:
Yeah, good point. I guess I just find it unlikely to actually pause on the same promise multiple times. I think the simplest thing would be to add a `.has` check before `.set` that throws an error if the promise is already tracked.